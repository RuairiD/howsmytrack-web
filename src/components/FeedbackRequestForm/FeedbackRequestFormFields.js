import React, { useState } from "react";

import { Button, Checkbox, Input, Form, Spin } from "antd";
import getFeedbackPromptPlaceholder from "./getFeedbackPromptPlaceholder";
import GenreSelectLabel from "./GenreSelectLabel";
import TracklessRadioGroup from "./TracklessRadioGroup";
import GenreSelect from "./GenreSelect";
import MissingGenreMessage from "./MissingGenreMessage";

const FeedbackRequestFormFields = ({
    form,
    mediaUrl,
    onMediaUrlChange,
    feedbackPrompt,
    genre,
    emailWhenGrouped,
    trackless,
    onSubmit,
    isLoading,
    submitted,
}) => {
    const [feedbackPromptPlaceholder] = useState(getFeedbackPromptPlaceholder());

    const isRequestTrackless = form.getFieldValue("trackless") === "trackless";

    let tracklessValue = "track";
    if (trackless) {
        tracklessValue = "trackless";
    }

    let emailWhenGroupedValue = emailWhenGrouped;
    if (emailWhenGroupedValue === undefined) {
        emailWhenGroupedValue = true;
    }

    const fieldDecorators = {
        mediaUrl: form.getFieldDecorator(
            "mediaUrl",
            {
                rules: [
                    {
                        // A media url isn't required if the user has declared that
                        // they are making a 'trackless' request i.e. only giving
                        // feedback, not receiving it.
                        required: isRequestTrackless,
                        message: "Please provide a track URL",
                    },
                ],
                initialValue: mediaUrl,
            },
        ),
        feedbackPrompt: form.getFieldDecorator(
            "feedbackPrompt",
            {
                initialValue: feedbackPrompt,
            },
        ),
        emailWhenGrouped: form.getFieldDecorator(
            "emailWhenGrouped",
            {
                valuePropName: "checked",
                initialValue: emailWhenGroupedValue,
            },
        ),
        genre: form.getFieldDecorator(
            "genre",
            {
                initialValue: genre,
                rules: [
                    {
                        required: true,
                        message: <MissingGenreMessage isRequestTrackless={isRequestTrackless} />,
                    },
                ],
            },
        ),
        trackless: form.getFieldDecorator(
            "trackless",
            {
                initialValue: tracklessValue,
            },
        ),
    };

    return (
        <Spin spinning={isLoading}>
            <Form onSubmit={onSubmit} className="hmt-form">
                <Form.Item>
                    <TracklessRadioGroup decorator={fieldDecorators.trackless} />
                </Form.Item>
                {!isRequestTrackless && (
                    <React.Fragment>
                        <Form.Item label="Soundcloud/Google Drive/Dropbox/OneDrive URL">
                            {fieldDecorators.mediaUrl(
                                <Input
                                    onChange={onMediaUrlChange}
                                    autoFocus
                                />,
                            )}
                        </Form.Item>
                        <Form.Item label="Is there anything you would like specific feedback on? (optional)">
                            {fieldDecorators.feedbackPrompt(
                                <Input.TextArea
                                    rows={4}
                                    placeholder={feedbackPromptPlaceholder}
                                    disabled={isRequestTrackless}
                                />,
                            )}
                        </Form.Item>
                    </React.Fragment>
                )}
                <Form.Item
                    colon={false}
                    label={<GenreSelectLabel />}
                >
                    <GenreSelect decorator={fieldDecorators.genre} />
                </Form.Item>
                <Form.Item>
                    {fieldDecorators.emailWhenGrouped(<Checkbox>Email me when this request is added to a group.</Checkbox>)}
                </Form.Item>
                <Form.Item>
                    <Button
                        block
                        type="primary"
                        htmlType="submit"
                        disabled={isLoading || submitted}
                    >
                        Submit Feedback Request
                    </Button>
                </Form.Item>
            </Form>
        </Spin>
    );
};

export default FeedbackRequestFormFields;
