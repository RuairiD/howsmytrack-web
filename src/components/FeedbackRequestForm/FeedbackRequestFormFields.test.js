import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import axios from "axios";
import { act } from "react-dom/test-utils";
import FeedbackRequestFormFields from "./FeedbackRequestFormFields";

jest.mock("axios");
jest.mock("./getFeedbackPromptPlaceholder", () => () => "placeholder text");

const mockForm = {
    getFieldDecorator: jest.fn(() => (c) => c),
    getFieldValue: jest.fn(() => "track"),
};

describe("FeedbackRequestFormFields", () => {
    beforeEach(() => {
        axios.post.mockRestore();
    });

    it("renders an empty form for submitting a request", () => {
        const wrapper = shallow(
            <FeedbackRequestFormFields
                form={mockForm}
                onMediaUrlChange={jest.fn()}
                onSubmit={jest.fn()}
                isLoading={false}
                submitted={false}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("skips mediaUrl and feedbackPrompt fields if request is trackless", () => {
        const wrapper = shallow(
            <FeedbackRequestFormFields
                form={mockForm}
                onMediaUrlChange={jest.fn()}
                onSubmit={jest.fn()}
                isLoading={false}
                submitted={false}
                trackless
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("disables the submit button if a submission request is still loading", () => {
        const wrapper = shallow(
            <FeedbackRequestFormFields
                form={mockForm}
                onMediaUrlChange={jest.fn()}
                onSubmit={jest.fn()}
                isLoading
                submitted={false}
                emailWhenGrouped
                trackless
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("calls the onMediaUrlChange callback when the mediaUrl is changed", () => {
        const onMediaUrlChange = jest.fn();
        const wrapper = mount(
            <FeedbackRequestFormFields
                form={mockForm}
                onMediaUrlChange={onMediaUrlChange}
                onSubmit={jest.fn()}
                isLoading={false}
                submitted={false}
            />,
        );

        act(() => wrapper.find("Input").get(0).props.onChange());

        expect(onMediaUrlChange).toHaveBeenCalled();
    });

    it("calls the onSubmit callback when the form is submitted, and shows any errors", () => {
        const onSubmit = jest.fn();
        const wrapper = mount(
            <FeedbackRequestFormFields
                form={mockForm}
                onMediaUrlChange={jest.fn()}
                onSubmit={onSubmit}
                isLoading={false}
                submitted={false}
            />,
        );

        act(() => wrapper.find("Form").get(0).props.onSubmit());

        expect(onSubmit).toHaveBeenCalled();
    });
});
