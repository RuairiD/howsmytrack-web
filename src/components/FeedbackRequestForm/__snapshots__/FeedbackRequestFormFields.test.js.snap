// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackRequestFormFields disables the submit button if a submission request is still loading 1`] = `
<Spin
  size="default"
  spinning={true}
  wrapperClassName=""
>
  <Form
    className="hmt-form"
    colon={true}
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[MockFunction]}
  >
    <FormItem
      hasFeedback={false}
    >
      <TracklessRadioGroup
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Soundcloud/Google Drive/Dropbox/OneDrive URL"
    >
      <Input
        autoFocus={true}
        onChange={[MockFunction]}
        type="text"
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Is there anything you would like specific feedback on? (optional)"
    >
      <TextArea
        disabled={false}
        placeholder="placeholder text"
        rows={4}
      />
    </FormItem>
    <FormItem
      colon={false}
      hasFeedback={false}
      label={<GenreSelectLabel />}
    >
      <GenreSelect
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Checkbox
        indeterminate={false}
      >
        Email me when this request is added to a group.
      </Checkbox>
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Button
        block={true}
        disabled={true}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Submit Feedback Request
      </Button>
    </FormItem>
  </Form>
</Spin>
`;

exports[`FeedbackRequestFormFields renders an empty form for submitting a request 1`] = `
<Spin
  size="default"
  spinning={false}
  wrapperClassName=""
>
  <Form
    className="hmt-form"
    colon={true}
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[MockFunction]}
  >
    <FormItem
      hasFeedback={false}
    >
      <TracklessRadioGroup
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Soundcloud/Google Drive/Dropbox/OneDrive URL"
    >
      <Input
        autoFocus={true}
        onChange={[MockFunction]}
        type="text"
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Is there anything you would like specific feedback on? (optional)"
    >
      <TextArea
        disabled={false}
        placeholder="placeholder text"
        rows={4}
      />
    </FormItem>
    <FormItem
      colon={false}
      hasFeedback={false}
      label={<GenreSelectLabel />}
    >
      <GenreSelect
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Checkbox
        indeterminate={false}
      >
        Email me when this request is added to a group.
      </Checkbox>
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Button
        block={true}
        disabled={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Submit Feedback Request
      </Button>
    </FormItem>
  </Form>
</Spin>
`;

exports[`FeedbackRequestFormFields skips mediaUrl and feedbackPrompt fields if request is trackless 1`] = `
<Spin
  size="default"
  spinning={false}
  wrapperClassName=""
>
  <Form
    className="hmt-form"
    colon={true}
    hideRequiredMark={false}
    layout="horizontal"
    onSubmit={[MockFunction]}
  >
    <FormItem
      hasFeedback={false}
    >
      <TracklessRadioGroup
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Soundcloud/Google Drive/Dropbox/OneDrive URL"
    >
      <Input
        autoFocus={true}
        onChange={[MockFunction]}
        type="text"
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Is there anything you would like specific feedback on? (optional)"
    >
      <TextArea
        disabled={false}
        placeholder="placeholder text"
        rows={4}
      />
    </FormItem>
    <FormItem
      colon={false}
      hasFeedback={false}
      label={<GenreSelectLabel />}
    >
      <GenreSelect
        decorator={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Checkbox
        indeterminate={false}
      >
        Email me when this request is added to a group.
      </Checkbox>
    </FormItem>
    <FormItem
      hasFeedback={false}
    >
      <Button
        block={true}
        disabled={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Submit Feedback Request
      </Button>
    </FormItem>
  </Form>
</Spin>
`;
